compaction/compaction_manager.cc

```diff
@@ -573,6 +573,59 @@ class major_compaction_task_executor : public compaction_task_executor, public m
     }
 };
 
+class splitting_compaction_task_executor final : public major_compaction_task_executor {
+    sstables::compaction_type_options::splitting _opt;
+public:
+    splitting_compaction_task_executor(compaction_manager& mgr,
+            throw_if_stopping do_throw_if_stopping,
+            table_state* t,
+            tasks::task_id parent_id,
+            sstables::compaction_type_options::splitting opt)
+        : major_compaction_task_executor(mgr, do_throw_if_stopping, t, parent_id)
+        , _opt(std::move(opt))
+    {
+    }
+private:
+    bool any_sstable_needs_split() const {
+        return bool(_compacting_table->main_sstable_set().for_each_sstable_until([this] (const sstables::shared_sstable& sst) {
+            bool needs_split = _opt.classifier(sst->get_first_decorated_key().token()) != _opt.classifier(sst->get_last_decorated_key().token());
+            return stop_iteration(needs_split);
+        }));
+    }
+protected:
+    sstring describe() const override {
+        return "Splitting compaction";
+    }
+
+    sstables::compaction_descriptor compaction_descriptor() const override {
+        auto desc = major_compaction_task_executor::compaction_descriptor();
+        desc.options = sstables::compaction_type_options::make_splitting(_opt.classifier);
+        return desc;
+    }
+
+    void on_compaction_start() override {
+        cmlog.info0("Splitting compaction started on behalf of {} and group {}", *_compacting_table, _compacting_table->get_group_id());
+    }
+
+    future<compaction_manager::compaction_stats_opt> do_run() override {
+        while (can_proceed() && any_sstable_needs_split()) {
+            std::exception_ptr ex;
+            try {
+                co_await major_compaction_task_executor::do_run();
```

Comment:
Shouldn't each sstable be split independently? This way if an sstable belongs to one half already (LCS, ICS), it need not be compacted.

This looks more like upgradesstables than major compaction.